Objects/stringlib/fastsearch.h

```diff
@@ -324,222 +359,402 @@ typedef struct STRINGLIB(_pre) {
+
+    // Direction Dependent
+    const Py_ssize_t s_stt = dir == 1 ? 0 : n - 1;
+    const Py_ssize_t p_stt = dir == 1 ? 0 : m - 1;
+    const Py_ssize_t p_end = dir == 1 ? m - 1 : 0;
+    const Py_ssize_t dir_m_m1 = reversed ? -m_m1 : m_m1;
+    const STRINGLIB_CHAR *const ss = s + s_stt + dir_m_m1;
+    const STRINGLIB_CHAR p_last = p[p_end];
+
+    // Indexers
+    Py_ssize_t i, j, ip, jp;
+
+    // Use Bloom for len(haystack) >= 10 * len(needle)
+    unsigned long mask = 0;
+    Py_ssize_t true_gap = 0;
+    Py_ssize_t j_stop = m;
+    if (n >= 30 * m) {
+        j_stop = m_m1;
+        true_gap = m;
+        // Note: true_gap("___aa") = 1
+        STRINGLIB_CHAR p_tmp;
+        STRINGLIB_BLOOM_ADD(mask, p_last);
+        for (j = 1; j < m; j++) {
+            jp = p_end + (reversed ? j : -j);
+            p_tmp = p[jp];
+            STRINGLIB_BLOOM_ADD(mask, p_tmp);
+            if (true_gap == m && p_tmp == p_last) {
+                true_gap = j;
+            }
+        }
+        LOG("Good Suffix True Gap: %ld\n", true_gap);
+    }
+
+    // Total cost of two-way initialization
+    // Horspool Calibration
+    const double hrs_lcost = 4.0;               // average loop cost
+    const double hrs_hcost = 0.4;               // false positive hit cost
+    // Two-Way Calibration
+    const double twy_icost = 3.0 * (double)m;   // total initialization cost
+    const double twy_lcost = 3.0;               // loop cost
+    // Temporary
+    double exp_hrs, exp_twy, ll;   // expected run times & loops left
+    Py_ssize_t j_off;              // offset from last to leftmost window index
+    Py_ssize_t shift;
+    STRINGLIB_CHAR s_last;
+    // Counters
+    Py_ssize_t count = 0;
+    Py_ssize_t iloop = 0, ihits_last = 0;
+    Py_ssize_t ihits = 0, iloop_last = 0;
+    // Loop
+    for (i = 0; i <= w;) {
+        iloop++;
+        ip = reversed ? -i : i;
+        s_last = ss[ip];
+        LOG2("Last window ch: %c\n", s_last);
+        if (true_gap) {
+            shift = 0;
+            if (s_last != p_last) {
+                if (i < w && !STRINGLIB_BLOOM(mask, ss[ip+dir])) {
+                    shift = m_p1;
+                }
+                else {
+                    shift = Py_MAX(table[s_last & TABLE_MASK], 1);
+                }
+            }
+        }
+        else {
+            shift = table[s_last & TABLE_MASK];
+        }
+        if (shift) {
+            LOG("Shift: %ld\n", shift);
+            i += shift;
+            continue;
+        }
+        // assert(s_last == p_last);                               // true_gap
+        // assert((s_last & TABLE_MASK) == (p_last & TABLE_MASK)); // else
+        j_off = ip - p_end;
+        for (j = 0; j < j_stop; j++) {
+            ihits++;
+            jp = p_stt + (reversed ? -j : j);
+            LOG2("Checking j=%ld: %c vs %c\n", j, ss[j_off + jp], p[jp]);
+            if (ss[j_off + jp] != p[jp]) {
+                break;
+            }
+        }
+        if (j == j_stop) {
```

Comment:
```suggestion

        if (j == j_stop) {
```

Just to separate the for-loop

Reply:
That if statement is very closely related to the loop. I would like to leave it as it is for these cases.